libvast/builtins/formats/json.cpp

```diff
@@ -212,97 +216,90 @@ auto parse_doc(const FieldValidator& validator,
   return {};
 }
 
-auto create_field_validator(bool has_selector)
-  -> std::function<caf::expected<void>(const detail::field_guard&)> {
-  if (has_selector) {
-    return [](const detail::field_guard& guard) -> caf::expected<void> {
-      if (guard.field_exists())
-        return {};
-      return caf::make_error(ec::parse_error, "field not present in a schema");
+auto create_field_validator(bool has_selector, bool infer_types)
+  -> std::function<bool(const detail::field_guard&)> {
+  if (has_selector and not infer_types) {
+    return [](const detail::field_guard& guard) {
+      return guard.field_exists();
     };
   }
-  return [](const detail::field_guard&) -> caf::expected<void> {
-    return caf::expected<void>{};
+  return [](const detail::field_guard&) {
+    return true;
   };
 }
```

Comment:
This is a form of dependency injection to steer the behaviour of document parser.
The validator returns true if it the field is correct and can be parsed and false otherwise.

So in case of --no-infer we only want to parse fields that were part of the schema. Others should return false